### ISSUE TITLE
How can I get the raw size or content-length of the response?

### ISSUE THREAD

**Issue by 10W8K** [puppeteer/puppeteer#667]
I mean some response is gzipped, I want to get the size, thanks

**Comment by aslushnikov** [comment]
@alexyan do I get it right that you need the encoded size?

**Comment by 10W8K** [comment]
that's right @aslushnikov

**Comment by aslushnikov** [comment]
An upstream bug that blocks this: [crbug.com/764946](https://bugs.chromium.org/p/chromium/issues/detail?id=764946)

**Comment by aslushnikov** [comment]
@alexyan one more clarification. Do you actually need the total response size, including http headers, that was transferred over the network, or you are specifically interested in the size of encoded response body?

**Comment by 10W8K** [comment]
Yes, I just want to get the size of encoded response body. For example, to check if the response is too large. @aslushnikov

**Comment by kdekooter** [comment]
This code works for me, not sure if it is the unzipped size though:

```js
let pageSize = 0

page.on('response', response => {
  response.buffer().then(buffer => {
    pageSize += buffer.length
})
```

**Comment by 10W8K** [comment]
thanks @kdekooter, but that is the unzipped size

**Comment by adamblok** [comment]
How about this issue? I need it too actually and the bug you've linked wasn't fixed: https://bugs.chromium.org/p/chromium/issues/detail?id=764946

**Comment by ithinkihaveacat** [comment]
The `content-length` header of the response might be helpful: if provided, it's supposed to be size/length of the response body as it travels across the network. (i.e. the amount of data transferred.) It's not always provided, and not guaranteed to be correct, though.

**Comment by ithinkihaveacat** [comment]
@aslushnikov Why is this blocked on an upstream bug? I'm just after the value that appears in the "Size" column of Devtools' Network panel. (If this is actually incorrect, and doesn't include the size of the headers, then I guess that would be good to fix, but would it be possible to simply export that value via the Puppeteer API first?)

![screenshot from 2017-10-31 18 05 03](https://user-images.githubusercontent.com/51244/32240825-41c96824-be66-11e7-84f8-e07c5db8d60f.png)

**Comment by vijeshv** [comment]
is it possible to get the total rspose size ?

**Comment by ithinkihaveacat** [comment]
@vijeshv I don't believe it's possible to get the total response size. I've written up my workaround [on Stack Overflow](https://stackoverflow.com/q/48263345/11543).

**Comment by vijeshv** [comment]
Thanks Michael for swift reply. Pls let me know if you find any workaround to compute compressed bytes.

Thanks
Vij

**Comment by vijeshv** [comment]
I have tried the work around to get uncompressed respose size as you suggested. But found  page not loading after opening browser. a kind of browser hanging behaviour. 
pls find script below.	

const browser = await puppeteer.launch({ headless:false,slowMo:250});
	 const page = await browser.newPage();
	 
	const p=await browser.pages();
	
const resources = {};

page.client.on('Network.dataReceived', (event) => {
  const request = page.networkManager.requestIdToRequest.get(event.requestId);
  if (request && request.url().startsWith('data:')) {
    return;
  }

  // encodedDataLength is supposed to be the amount of data received
  // over the wire, but it's often 0, so just use dataLength for consistency.
  // https://chromedevtools.github.io/devtools-protocol/tot/Network/#event-dataReceived
  // const length = event.encodedDataLength > 0 ?
  //     event.encodedDataLength : event.dataLength;
  const length = event.dataLength;
  if (url in resources) {
    resources[url] += length;
  } else {
    resources[url] = length;
  }
  console.log("length:"+length);
  	
const totalUncompressedBytes = Object.values(resources).reduce((a, n) => a + n);	
console.log("uncompressed byte :" + totalUncompressedBytes);
});



   await page.goto('https://www.nytimes.com/', { timeout: 50000 });
await browser.close();

**Comment by Mr0grog** [comment]
Going with the blocking issue above (crbug.com/764946) I was looking at working around this by checking the length of `Response.headersText`:

```js
const devToolsResponses = new Map();
const devTools = await page.target().createCDPSession();
await devTools.send('Network.enable');

devTools.on('Network.responseReceived', event => {
  devToolsResponses.set(event.requestId, event.response);
});

devTools.on('Network.loadingFinished', event => {
  const response = devToolsResponses.get(event.requestId);
  const encodedBodyLength = event.encodedDataLength - response.headersText.length;
  console.log(`${encodedBodyLength} bytes for ${response.url}`);
});
```

…but it turns out that `response.headersText` is `undefined` for HTTP/2 responses. I *think* that makes sense because HTTP/2 headers are binary, so there’s no textual representation, but then my question is: are headers not included in the `encodedDataLength` for HTTP/2 (since they are also a separate, special frame) or do I need to be doing something to estimate how many bytes they were?

**Comment by JohnBruce** [comment]
@aslushnikov Is this still blocked by https://bugs.chromium.org/p/chromium/issues/detail?id=764946 ?  Looks like that got marked as wontfix nearly a year ago.

**Comment by aslushnikov** [comment]
>  Is this still blocked by https://bugs.chromium.org/p/chromium/issues/detail?id=764946 ? 

@JohnBruce no; we still need to do work upstream though to report the zipped body size.

**Comment by Sinequanonh** [comment]
@aslushnikov that'd be awesome 🙏Do you have an ETA on it?

**Comment by nylen** [comment]
> This code works for me, not sure if it is the unzipped size though:
> 
> ```js
> let pageSize = 0
> 
> page.on('response', response => {
>   response.buffer().then(buffer => {
>     pageSize += buffer.length
> })
> ```

There are a couple of issues with this code:

1. If the `page` object is re-used for many page loads, then this will add many event listeners over time
2. Some responses like redirects will cause `response.buffer()` to fail

Here is an updated version:

```js
let pageTotalBytes = 0;

async function addResponseSize( response ) { 
    try {
        const buffer = await response.buffer();
        pageTotalBytes += buffer.length;
    } catch {
        // Error: Response body is unavailable for redirect responses
        // TODO: other errors possible?
    }
}

// Add the event handler
page.on( 'response', addResponseSize );

// Load the page
const res = await page.goto( yourUrl );
// and maybe do other operations you want to measure
// like clicks that might cause AJAX requests

// Clean up: remove the event handler
page.off( 'response', addResponseSize );

// The pageTotalBytes variable is now ready for use
```
